<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Tutorial &mdash; featurExtract 0.1 documentation</title><link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/copybutton.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script src="_static/clipboard.min.js"></script>
        <script src="_static/copybutton.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Example" href="example.html" />
    <link rel="prev" title="Installation" href="install.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> featurExtract
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#quick-start">Quick-start</a></li>
<li class="toctree-l2"><a class="reference internal" href="#main-functional-commands">Main functional commands</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#create">Create</a></li>
<li class="toctree-l3"><a class="reference internal" href="#promoter">Promoter</a></li>
<li class="toctree-l3"><a class="reference internal" href="#utr">UTR</a></li>
<li class="toctree-l3"><a class="reference internal" href="#uorf">uORF</a></li>
<li class="toctree-l3"><a class="reference internal" href="#cds">CDS</a></li>
<li class="toctree-l3"><a class="reference internal" href="#dorf">dORF</a></li>
<li class="toctree-l3"><a class="reference internal" href="#cdna-mrna">cDNA/mRNA</a></li>
<li class="toctree-l3"><a class="reference internal" href="#gene">Gene</a></li>
<li class="toctree-l3"><a class="reference internal" href="#exon">Exon</a></li>
<li class="toctree-l3"><a class="reference internal" href="#intron">Intron</a></li>
<li class="toctree-l3"><a class="reference internal" href="#igr">IGR</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="example.html">Example</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">FAQ</a></li>
<li class="toctree-l1"><a class="reference internal" href="about.html">About</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">featurExtract</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Tutorial</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/tutorial.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="section" id="tutorial">
<h1>Tutorial<a class="headerlink" href="#tutorial" title="Permalink to this headline">¶</a></h1>
<p>This tutorial introduces the text-based usage and workflow in featurExtract.
featurExtract is designed for GFF and GTF file, meanwhile GenBankExtract is suited for GenBank file.</p>
<p><cite>featurExtract</cite> is <a class="reference external" href="https://github.com/SitaoZ/featurExtract">tested</a> with Python 3.8</p>
<div class="section" id="quick-start">
<h2>Quick-start<a class="headerlink" href="#quick-start" title="Permalink to this headline">¶</a></h2>
<p>Find featurextract in command line and show the commands provided.</p>
<div class="highlight-bash notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24</pre></div></td><td class="code"><div class="highlight"><pre><span></span><span class="c1"># find the featurExtract exec path</span>
which featurExtract
/opt/anaconda3/bin/featurExtract
<span class="c1"># show help info</span>
featurExtract -h
<span class="c1">#    Program:  featurExtract (pipeline for genomic feature extract)</span>
<span class="c1">#    Version:  0.2.5.3</span>
<span class="c1">#    Contact:  Sitao Zhu &lt;zhusitao1990@163.com&gt;</span>
<span class="c1">#    Usage  :  featurExtract &lt;command&gt; [parameters]</span>
<span class="c1">#    Command:</span>
<span class="c1">#        create        create database for GTF or GFF</span>
<span class="c1">#        coverage      read coverage by transcript models</span>
<span class="c1">#        CDS           extract CDS for gene or genome</span>
<span class="c1">#        cdna          extract cdna (refMrna) sequence fo gene or genome</span>
<span class="c1">#        dORF          extract downstream (dORF) for gene or genome</span>
<span class="c1">#        exon          extract exon for transcript</span>
<span class="c1">#        gene          extract gene sequence fo gene or genome</span>
<span class="c1">#        intron        extract intron for transcript</span>
<span class="c1">#        IGR           extract intergenic region (IGR) between genes</span>
<span class="c1">#        mRNA          extract mature messager RNA for transcript</span>
<span class="c1">#        promoter      extract promoter for gene or genome</span>
<span class="c1">#        terminator    extract terminator for gene or genome</span>
<span class="c1">#        uORF          extract upstream ORF (uORF) for gene or genome</span>
<span class="c1">#        UTR           extract untranslated region (UTR) for gene or genome</span>
</pre></div>
</td></tr></table></div>
</div>
<div class="section" id="main-functional-commands">
<h2>Main functional commands<a class="headerlink" href="#main-functional-commands" title="Permalink to this headline">¶</a></h2>
<p>In this section, we will introduce commands provided by featurExtract modual.
How to get help info, and how to use them. Subcommands are show helpdoc when
the specific command is provided as args in command line.</p>
<div class="section" id="create">
<h3>Create<a class="headerlink" href="#create" title="Permalink to this headline">¶</a></h3>
<p>The create command use <a class="reference external" href="https://github.com/daler/gffutils">gffutils</a> to build a sqlite database, which
stores the genome and gene structure need for other functional commands in featurExtract.</p>
<div class="highlight-bash notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15</pre></div></td><td class="code"><div class="highlight"><pre><span></span>featurExtract create -h
<span class="c1">#    usage: featurExtract create [-h] [-f {GFF,GTF}] -g GENOMEFEATURE -o OUTPUT_PREFIX</span>
<span class="c1">#    optional arguments:</span>
<span class="c1">#        -h, --help            show this help message and exit</span>
<span class="c1">#        -f {GFF,GTF}, --file_type {GFF,GTF}</span>
<span class="c1">#                      genome annotation file</span>
<span class="c1">#        -g GENOMEFEATURE, --genomefeature GENOMEFEATURE</span>
<span class="c1">#                      genome annotation file</span>
<span class="c1">#        -o OUTPUT_PREFIX, --output_prefix OUTPUT_PREFIX</span>
<span class="c1">#                      database absolute path</span>

<span class="c1"># GFF</span>
featurExtract create -f GFF -g ath.gff3 -o ath
<span class="c1"># or GTF</span>
featurExtract create -f GTF -g ath.gtf  -o ath
</pre></div>
</td></tr></table></div>
</div>
<div class="section" id="promoter">
<h3>Promoter<a class="headerlink" href="#promoter" title="Permalink to this headline">¶</a></h3>
<p>Promoter sequence is essential for gene functional resaerch. Usually.</p>
<div class="highlight-bash notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25</pre></div></td><td class="code"><div class="highlight"><pre><span></span>featurExtract promoter -h
<span class="c1">#usage: featurExtract promoter [-h] -d DATABASE -f GENOME [-g GENE]</span>
<span class="c1">#                           [-l PROMOTER_LENGTH] [-u UTR5_UPPER_LENGTH]</span>
<span class="c1">#                           [-o OUTPUT] [-t {csv,fasta}] [-p]</span>
<span class="c1">#optional arguments:</span>
<span class="c1">#  -h, --help            show this help message and exit</span>
<span class="c1">#  -d DATABASE, --database DATABASE</span>
<span class="c1">#                        database generated by subcommand create</span>
<span class="c1">#  -f GENOME, --genome GENOME</span>
<span class="c1">#                        genome fasta path</span>
<span class="c1">#  -g GENE, --gene GENE  specific gene; if not given, return whole genes</span>
<span class="c1">#  -l PROMOTER_LENGTH, --promoter_length PROMOTER_LENGTH</span>
<span class="c1">#                        promoter length before TSS (default 100 nt)</span>
<span class="c1">#  -u UTR5_UPPER_LENGTH, --utr5_upper_length UTR5_UPPER_LENGTH</span>
<span class="c1">#                        5&#39; utr length after TSS (default 10 nt)</span>
<span class="c1">#  -o OUTPUT, --output OUTPUT</span>
<span class="c1">#                        output file path</span>
<span class="c1">#  -t {csv,fasta}, --output_format {csv,fasta}</span>
<span class="c1">#                        output format</span>
<span class="c1">#  -p, --print           output to stdout</span>

<span class="c1"># All genes in whole gnome</span>
featurExtract promoter -d ath.GFF -f ath.fa -l <span class="m">200</span> -u <span class="m">100</span> -o promoter.csv --output_format fasta
<span class="c1"># A given gene</span>
featurExtract promoter -d ath.GFF -f ath.fa -l <span class="m">200</span> -u <span class="m">100</span> -g AT1G01010 -p --output_format fasta
</pre></div>
</td></tr></table></div>
</div>
<div class="section" id="utr">
<h3>UTR<a class="headerlink" href="#utr" title="Permalink to this headline">¶</a></h3>
<p>UTR (untranslated region) sequence is essential for gene functional resaerch. In molecular genetics,
an UTR refers to either of two sections, one on each side of a coding sequence on a strand of mRNA.
If it is found on the 5’ side, it is called the 5’ UTR (or leader sequence), or if it is found on
the 3’ side, it is called the 3’ UTR (or trailer sequence).</p>
<div class="highlight-bash notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23</pre></div></td><td class="code"><div class="highlight"><pre><span></span>featurExtract UTR -h
<span class="c1">#usage: featurExtract UTR [-h] -d DATABASE -f GENOME [-i TRANSCRIPT]</span>
<span class="c1">#                     [-o OUTPUT] [-p] [-s {GFF,GTF}]</span>
<span class="c1">#</span>
<span class="c1">#optional arguments:</span>
<span class="c1">#  -h, --help            show this help message and exit</span>
<span class="c1">#  -d DATABASE, --database DATABASE</span>
<span class="c1">#                        database generated by subcommand create</span>
<span class="c1">#  -f GENOME, --genome GENOME</span>
<span class="c1">#                        genome fasta file</span>
<span class="c1">#  -i TRANSCRIPT, --transcript TRANSCRIPT</span>
<span class="c1">#                        specific transcript id; if not given, whole transcript</span>
<span class="c1">#                        will return</span>
<span class="c1">#  -o OUTPUT, --output OUTPUT</span>
<span class="c1">#                        output file path</span>
<span class="c1">#  -p, --print           output to stdout</span>
<span class="c1">#  -s {GFF,GTF}, --style {GFF,GTF}</span>
<span class="c1">#                        GTF database or GFF database</span>
<span class="c1">#</span>
<span class="c1"># All transcripts in whole gnome</span>
featurExtract promoter -d ath.GFF -f ath.fa -o utr.csv -s GFF
<span class="c1"># A given gene</span>
featurExtract promoter -d ath.GFF -f ath.fa -i AT1G01010.1 -p -s GFF
</pre></div>
</td></tr></table></div>
</div>
<div class="section" id="uorf">
<h3>uORF<a class="headerlink" href="#uorf" title="Permalink to this headline">¶</a></h3>
<p>uORF (upstream open reading frame), is an open reading frame (ORF) within the 5’ untranslated region (5’UTR) of an mRNA.
uORFs can regulate eukaryotic gene expression and repress downstream expression of the primary ORF.</p>
<div class="highlight-bash notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27</pre></div></td><td class="code"><div class="highlight"><pre><span></span>featurExtract uORF -h

<span class="c1">#usage: featurExtract uORF [-h] -d DATABASE -f GENOME [-i TRANSCRIPT]</span>
<span class="c1">#                      [-t {csv,fasta,gff}] [-o OUTPUT] [-m] [-n]</span>
<span class="c1">#                      [-s {GFF,GTF}]</span>

<span class="c1">#optional arguments:</span>
<span class="c1">#  -h, --help            show this help message and exit</span>
<span class="c1">#  -d DATABASE, --database DATABASE</span>
<span class="c1">#                        database generated by subcommand create</span>
<span class="c1">#  -f GENOME, --genome GENOME</span>
<span class="c1">#                        genome fasta</span>
<span class="c1">#  -i TRANSCRIPT, --transcript TRANSCRIPT</span>
<span class="c1">#                        specific transcript id; if not given, whole transcript</span>
<span class="c1">#                        will return</span>
<span class="c1">#  -t {csv,fasta,gff}, --output_format {csv,fasta,gff}</span>
<span class="c1">#                        output format</span>
<span class="c1">#  -o OUTPUT, --output OUTPUT</span>
<span class="c1">#                        output file path</span>
<span class="c1">#  -m, --schematic_without_intron</span>
<span class="c1">#                        schematic figure file for uORF, CDS and transcript</span>
<span class="c1">#                        without intron</span>
<span class="c1">#  -n, --schematic_with_intron</span>
<span class="c1">#                        schematic figure file for uORF, CDS and transcript</span>
<span class="c1">#                        with intron</span>
<span class="c1">#  -s {GFF,GTF}, --style {GFF,GTF}</span>
<span class="c1">#                        GTF database or GFF database</span>
</pre></div>
</td></tr></table></div>
</div>
<div class="section" id="cds">
<h3>CDS<a class="headerlink" href="#cds" title="Permalink to this headline">¶</a></h3>
<p>The CDS (coding sequence), is the portion of a gene’s DNA or RNA that codes for protein.</p>
<div class="highlight-bash notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27</pre></div></td><td class="code"><div class="highlight"><pre><span></span>featurExtract CDS -h

<span class="c1">#usage: featurExtract uORF [-h] -d DATABASE -f GENOME [-i TRANSCRIPT]</span>
<span class="c1">#                      [-t {csv,fasta,gff}] [-o OUTPUT] [-m] [-n]</span>
<span class="c1">#                      [-s {GFF,GTF}]</span>

<span class="c1">#optional arguments:</span>
<span class="c1">#  -h, --help            show this help message and exit</span>
<span class="c1">#  -d DATABASE, --database DATABASE</span>
<span class="c1">#                        database generated by subcommand create</span>
<span class="c1">#  -f GENOME, --genome GENOME</span>
<span class="c1">#                        genome fasta</span>
<span class="c1">#  -i TRANSCRIPT, --transcript TRANSCRIPT</span>
<span class="c1">#                        specific transcript id; if not given, whole transcript</span>
<span class="c1">#                        will return</span>
<span class="c1">#  -t {csv,fasta,gff}, --output_format {csv,fasta,gff}</span>
<span class="c1">#                        output format</span>
<span class="c1">#  -o OUTPUT, --output OUTPUT</span>
<span class="c1">#                        output file path</span>
<span class="c1">#  -m, --schematic_without_intron</span>
<span class="c1">#                        schematic figure file for uORF, CDS and transcript</span>
<span class="c1">#                        without intron</span>
<span class="c1">#  -n, --schematic_with_intron</span>
<span class="c1">#                        schematic figure file for uORF, CDS and transcript</span>
<span class="c1">#                        with intron</span>
<span class="c1">#  -s {GFF,GTF}, --style {GFF,GTF}</span>
<span class="c1">#                        GTF database or GFF database</span>
</pre></div>
</td></tr></table></div>
</div>
<div class="section" id="dorf">
<h3>dORF<a class="headerlink" href="#dorf" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27</pre></div></td><td class="code"><div class="highlight"><pre><span></span>featurExtract dORF -h
<span class="c1">#usage: featurExtract dORF [-h] -d DATABASE -f GENOME [-i TRANSCRIPT]</span>
<span class="c1">#                      [-t {csv,fasta,gff}] [-o OUTPUT]</span>
<span class="c1">#                      [-m SCHEMATIC_WITHOUT_INTRON]</span>
<span class="c1">#                      [-n SCHEMATIC_WITH_INTRON] [-s {GFF,GTF}]</span>

<span class="c1">#optional arguments:</span>
<span class="c1">#  -h, --help            show this help message and exit</span>
<span class="c1">#  -d DATABASE, --database DATABASE</span>
<span class="c1">#                        database generated by subcommand create</span>
<span class="c1">#  -f GENOME, --genome GENOME</span>
<span class="c1">#                        genome fasta</span>
<span class="c1">#  -i TRANSCRIPT, --transcript TRANSCRIPT</span>
<span class="c1">#                        specific transcript id; if not given, whole transcript</span>
<span class="c1">#                        will return</span>
<span class="c1">#  -t {csv,fasta,gff}, --output_format {csv,fasta,gff}</span>
<span class="c1">#                        output format</span>
<span class="c1">#  -o OUTPUT, --output OUTPUT</span>
<span class="c1">#                        output file path</span>
<span class="c1">#  -m SCHEMATIC_WITHOUT_INTRON, --schematic_without_intron SCHEMATIC_WITHOUT_INTRON</span>
<span class="c1">#                        schematic figure file for dORF, CDS and transcript</span>
<span class="c1">#                        without intron</span>
<span class="c1">#  -n SCHEMATIC_WITH_INTRON, --schematic_with_intron SCHEMATIC_WITH_INTRON</span>
<span class="c1">#                        schematic figure file for dORF, CDS and transcript</span>
<span class="c1">#                        with intron</span>
<span class="c1">#  -s {GFF,GTF}, --style {GFF,GTF}</span>
<span class="c1">#                        GTF database or GFF database</span>
</pre></div>
</td></tr></table></div>
</div>
<div class="section" id="cdna-mrna">
<h3>cDNA/mRNA<a class="headerlink" href="#cdna-mrna" title="Permalink to this headline">¶</a></h3>
<p>In genetics, complementary DNA (cDNA) is DNA synthesized from a single-stranded RNA (e.g., messenger RNA (mRNA) or microRNA (miRNA)) template in a reaction catalyzed by the enzyme reverse transcriptase.</p>
<div class="highlight-bash notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22</pre></div></td><td class="code"><div class="highlight"><pre><span></span>featurExtract cdna -h
<span class="c1">#usage: featurExtract cdna [-h] -d DATABASE -f GENOME [-i TRANSCRIPT]</span>
<span class="c1">#                      [-o OUTPUT] [-t {csv,fasta}] [-p] [-u]</span>
<span class="c1">#                      [-s {GFF,GTF}]</span>

<span class="c1">#optional arguments:</span>
<span class="c1">#  -h, --help            show this help message and exit</span>
<span class="c1">#  -d DATABASE, --database DATABASE</span>
<span class="c1">#                        database generated by subcommand create</span>
<span class="c1">#  -f GENOME, --genome GENOME</span>
<span class="c1">#                        genome fasta</span>
<span class="c1">#  -i TRANSCRIPT, --transcript TRANSCRIPT</span>
<span class="c1">#                        specific transcript; if not given, return whole</span>
<span class="c1">#                        transcripts</span>
<span class="c1">#  -o OUTPUT, --output OUTPUT</span>
<span class="c1">#                        output file path</span>
<span class="c1">#  -t {csv,fasta}, --output_format {csv,fasta}</span>
<span class="c1">#                        output format</span>
<span class="c1">#  -p, --print           output to stdout</span>
<span class="c1">#  -u, --upper           upper CDS and lower utr</span>
<span class="c1">#  -s {GFF,GTF}, --style {GFF,GTF}</span>
<span class="c1">#                        GTF database or GFF database</span>
</pre></div>
</td></tr></table></div>
</div>
<div class="section" id="gene">
<h3>Gene<a class="headerlink" href="#gene" title="Permalink to this headline">¶</a></h3>
<p>gene</p>
<div class="highlight-bash notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14</pre></div></td><td class="code"><div class="highlight"><pre><span></span>featurExtract gene -h
<span class="c1">#usage: featurExtract gene [-h] -d DATABASE -f GENOME [-g GENE] [-o OUTPUT]</span>
<span class="c1">#                      [-p]</span>

<span class="c1">#optional arguments:</span>
<span class="c1">#  -h, --help            show this help message and exit</span>
<span class="c1">#  -d DATABASE, --database DATABASE</span>
<span class="c1">#                        database generated by subcommand create</span>
<span class="c1">#  -f GENOME, --genome GENOME</span>
<span class="c1">#                        genome fasta</span>
<span class="c1">#  -g GENE, --gene GENE  specific gene; if not given, return whole genes</span>
<span class="c1">#  -o OUTPUT, --output OUTPUT</span>
<span class="c1">#                        output file path</span>
<span class="c1">#  -p, --print           output to stdout</span>
</pre></div>
</td></tr></table></div>
</div>
<div class="section" id="exon">
<h3>Exon<a class="headerlink" href="#exon" title="Permalink to this headline">¶</a></h3>
<p>An exon is any part of a gene that will form a part of the final mature RNA produced by that gene after introns have been removed by RNA splicing.</p>
<div class="highlight-bash notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17</pre></div></td><td class="code"><div class="highlight"><pre><span></span>featurExtract exon -h
<span class="c1">#usage: featurExtract exon [-h] -d DATABASE -f GENOME [-i TRANSCRIPT]</span>
<span class="c1">#                      [-o OUTPUT] [-p] [-s {GFF,GTF}]</span>

<span class="c1">#optional arguments:</span>
<span class="c1">#  -h, --help            show this help message and exit</span>
<span class="c1">#  -d DATABASE, --database DATABASE</span>
<span class="c1">#                        database generated by subcommand create</span>
<span class="c1">#  -f GENOME, --genome GENOME</span>
<span class="c1">#                        genome fasta</span>
<span class="c1">#  -i TRANSCRIPT, --transcript TRANSCRIPT</span>
<span class="c1">#                        specific transcript id; needed</span>
<span class="c1">#  -o OUTPUT, --output OUTPUT</span>
<span class="c1">#                        output file path</span>
<span class="c1">#  -p, --print           output to stdout</span>
<span class="c1">#  -s {GFF,GTF}, --style {GFF,GTF}</span>
<span class="c1">#                        GTF database or GFF database</span>
</pre></div>
</td></tr></table></div>
</div>
<div class="section" id="intron">
<h3>Intron<a class="headerlink" href="#intron" title="Permalink to this headline">¶</a></h3>
<p>An intron is any nucleotide sequence within a gene that is removed by RNA processing during production of the final RNA product.</p>
<div class="highlight-bash notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17</pre></div></td><td class="code"><div class="highlight"><pre><span></span>featurExtract intron -h
<span class="c1">#usage: featurExtract intron [-h] -d DATABASE -f GENOME [-i TRANSCRIPT]</span>
<span class="c1">#                        [-o OUTPUT] [-p] [-s {GFF,GTF}]</span>

<span class="c1">#optional arguments:</span>
<span class="c1">#  -h, --help            show this help message and exit</span>
<span class="c1">#  -d DATABASE, --database DATABASE</span>
<span class="c1">#                        database generated by subcommand create</span>
<span class="c1">#  -f GENOME, --genome GENOME</span>
<span class="c1">#                        genome fasta</span>
<span class="c1">#  -i TRANSCRIPT, --transcript TRANSCRIPT</span>
<span class="c1">#                        specific transcript id; needed</span>
<span class="c1">#  -o OUTPUT, --output OUTPUT</span>
<span class="c1">#                        output file path</span>
<span class="c1">#  -p, --print           output to stdout</span>
<span class="c1">#  -s {GFF,GTF}, --style {GFF,GTF}</span>
<span class="c1">#                        GTF database or GFF database</span>
</pre></div>
</td></tr></table></div>
</div>
<div class="section" id="igr">
<h3>IGR<a class="headerlink" href="#igr" title="Permalink to this headline">¶</a></h3>
<p>An IGR (intergenic region) is a stretch of DNA sequences located between genes.</p>
<div class="highlight-bash notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18</pre></div></td><td class="code"><div class="highlight"><pre><span></span>featurExtract IGR -h
<span class="c1">#usage: featurExtract IGR [-h] -d DATABASE -f GENOME [-l IGR_LENGTH]</span>
<span class="c1">#                         [-o OUTPUT] [-p] [-s {GFF,GTF}]</span>

<span class="c1">#optional arguments:</span>
<span class="c1">#  -h, --help            show this help message and exit</span>
<span class="c1">#  -d DATABASE, --database DATABASE</span>
<span class="c1">#                        database generated by subcommand create</span>
<span class="c1">#  -f GENOME, --genome GENOME</span>
<span class="c1">#                        genome fasta</span>
<span class="c1">#  -l IGR_LENGTH, --IGR_length IGR_LENGTH</span>
<span class="c1">#                        IGR length threshold</span>
<span class="c1">#  -o OUTPUT, --output OUTPUT</span>
<span class="c1">#                        output fasta file path</span>
<span class="c1">#  -p, --print           output to stdout</span>
<span class="c1">#  -s {GFF,GTF}, --style {GFF,GTF}</span>
<span class="c1">#                        GTF database only contain protein genes, while GFF</span>
<span class="c1">#                        database contain protein genes and nocoding genes</span>
</pre></div>
</td></tr></table></div>
<ul class="simple">
<li><p>python &gt;= 3.7.6 <a class="reference external" href="https://www.python.org/">python</a></p></li>
<li><p>pandas &gt;= 1.2.4 <a class="reference external" href="https://pandas.pydata.org/docs/">pandas</a></p></li>
<li><p>gffutils &gt;= 0.10.1 <a class="reference external" href="https://pythonhosted.org/gffutils/">gffutils</a></p></li>
<li><p>setuptools &gt;= 49.2.0 <a class="reference external" href="https://pypi.org/project/setuptools/">setuptools</a></p></li>
<li><p>BioPython &gt;= 1.78 <a class="reference external" href="https://biopython.org/wiki/Documentation/">biopython</a></p></li>
</ul>
<p>Install them all with <cite>conda</cite>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">install</span> <span class="o">--</span><span class="n">channel</span> <span class="n">conda</span><span class="o">-</span><span class="n">forge</span> <span class="o">--</span><span class="n">channel</span> <span class="n">python</span> <span class="n">pandas</span> <span class="n">gffutils</span> <span class="n">setuptools</span> <span class="n">biopython</span>
</pre></div>
</div>
</div>
</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="install.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="example.html" class="btn btn-neutral float-right" title="Example" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, zhusitao.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>